File: leetcode/problems/cpp/k-closest-points-to-origin.cpp
// by @codeAbinash
// Time : O(n log k)
// Space : O(k)

#include "queue"
#include "vector"

using namespace std;

class Solution {
 public:
   vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
      priority_queue<pair<int, int>> pq;
      for (int i = 0; i < points.size(); i++) {
         int x = points[i][0], y = points[i][1];
         pq.push({x * x + y * y, i});
         if (pq.size() > k) pq.pop();
      }
      vector<vector<int>> ans;
      while (!pq.empty()) {
         ans.push_back(points[pq.top().second]);
         pq.pop();
      }
      return ans;
   }
};

File: leetcode/problems/java/k-closest-points-to-origin.java
import java.util.PriorityQueue;

class Solution {
  public int[][] kClosest(int[][] points, int k) {
    PriorityQueue<int[]> maxHeap = new PriorityQueue<>((a, b) -> b[0] * b[0] + b[1] * b[1] - a[0] * a[0] - a[1] * a[1]);
    for (int[] point : points) {
      maxHeap.add(point);
      if (maxHeap.size() > k) {
        maxHeap.poll();
      }
    }
    int[][] result = new int[k][2];
    while (k-- > 0) {
      result[k] = maxHeap.poll();
    }
    return result;
  }
}